### ISSUE TITLE
New Salt server slow to sync cache from S3

### ISSUE THREAD

**Issue by richardgavel** [saltstack/salt#48641]
### Description of Issue/Question

I am using a Docker container as a test bed for our existing Saltstack server, which uses a s3fs. However, upon startup, it has to sync the local cache from S3, and it takes forever due to the number of files (over 5 GB worth). One issue is that the s3fs pulls down the files one at a time. Can it do it in parallel? Also, can it wait to populate a file from S3 locally until it is requested (or at least have a flag to defer local cache population)?

**Comment by terminalmage** [comment]
Not off the top of my head, though I know that @basepi has semi-recently made some improvements to s3fs, I _think_ for this very reason. Am I remembering correctly, @basepi?

**Comment by basepi** [comment]
I did not do any performance improvements. Our use in hubble only updates the cache once per hour so speed wasn't super important to us.

The primary improvement I made was to make ordering explicit. It was using dicts to store the cached data so fileserver ordering wasn't necessarily preserved (which is terrible if you're relying on fileserver ordering).

JIT caching is already "supported", but it's currently hardcoded as off. Would be pretty trivial to tie it to `__opts__`: https://github.com/saltstack/salt/blob/14ab5e79d433df674ab41e81726a24fb293022d4/salt/fileserver/s3fs.py#L107
